### ISSUE TITLE
Fails to start on GPT partitioned device with "lsblk: : not a block device"

### ISSUE THREAD

**Issue by jwmullally** [balena-os/balena-supervisor#2150]
The balena_supervisor container continuously crashes on our x86_64 GPT formatted systems, and we can no longer update them via the web console. We upgraded our supervisor version today.

```
root@fd36690:~# balena ps -a
CONTAINER ID   IMAGE                                                            COMMAND                  CREATED       STATUS                         PORTS     NAMES
d2f64407d83a   registry2.balena-cloud.com/v2/7c090259a71e8a88f569c09a5402e8fe   "/usr/src/app/entry.…"   4 hours ago   Restarting (32) 1 second ago             balena_supervisor
4298273b9431   9060b38cc6df                                                     "/usr/bin/entry.sh p…"   4 hours ago   Up 26 minutes                            main_6350798_2534313_e164482a9e5b9e89e547c3f1122543e0
```

```
root@fd36690:~# balena logs d2f64407d83a 2>&1 | tail -n 5
lsblk: : not a block device
lsblk: : not a block device
lsblk: : not a block device
lsblk: : not a block device
lsblk: : not a block device
```

```
root@fd36690:~# balena run --privileged --rm -it --entrypoint=/bin/sh registry2.balena-cloud.com/v2/7c090259a71e8a88f569c09a5402e8fe
/usr/src/app # ls
dist                 entry.sh             mount-partitions.sh  node_modules         package.json
/usr/src/app # ./entry.sh 
lsblk: : not a block device
```

```
/usr/src/app # ./mount-partitions.sh 
lsblk: : not a block device
ERROR: Could not determine boot device. Please launch Supervisor as a privileged container with host networking.
```

```
/usr/src/app # fdisk -l
Found valid GPT with protective MBR; using GPT

Disk /dev/mmcblk0: 241664000 sectors, 3312M
Logical sector size: 512
Disk identifier (GUID): 704f7515-8a11-4107-951b-2676dc541a27
Partition table holds up to 128 entries
First usable sector is 34, last usable sector is 241663966

Number  Start (sector)    End (sector)  Size Name
     1            8192           90111 40.0M resin-boot
     2           90112         2138111 1000M resin-rootA
     3         2138112         4186111 1000M resin-rootB
     4         4186112         4227071 20.0M resin-state
     5         4227072       241663966  113G resin-data
```

The code that seems to be raising the issue:
(introduced in https://github.com/balena-os/balena-supervisor/commit/49ee1042a87ce0b101054e32b84afaa324a371de ?)

```
/usr/src/app # cat mount-partitions.sh 
#!/bin/sh

# Mounts boot, state, & data partitions from balenaOS.
# The container must be privileged for this to function correctly.

# Get the current boot block device in case there are duplicate partition labels
# for `(balena|resin)-(boot|state|data)` found.
current_boot_block_device=""
if [ "${TEST}" != 1 ]; then
    current_boot_partition=$(fdisk -l | grep '*  ' | cut -d' ' -f1 2>&1)
    current_boot_block_device=$(lsblk -no pkname "${current_boot_partition}")
    if [ "${current_boot_block_device}" = "" ]; then
        echo "ERROR: Could not determine boot device. Please launch Supervisor as a privileged container with host networking."
        exit 1
    fi
fi

```

---

Some observations:

* entry.sh runs with `set -e` then sources `mount-partitions.sh`, which is why the script aborts before the error can be printed.
* GPT partition tables have no active partition that `fdisk -l | grep '*  '` would find.

**Comment by jwmullally** [comment]
Workaround fix that got the supervisor back working OK for us.

You will need to replace /dev/mmcblk0 below with your own block device. (I used `-` for the sed delimiter, so if your block device path includes that character, you'll have to change it something else).

```sh
#!/bin/sh
set -ex
# Workaround for https://github.com/balena-os/balena-supervisor/issues/2150
# balena-supervisor: Fails to start on GPT partitioned device with "lsblk: : not a block device"

SUP_ID=$(balena ps --all --quiet --no-trunc -f "name=balena_supervisor")
if [ -z "$SUP_ID" ]; then
    echo "No container found"
    exit 1
fi
systemctl stop balena || true
sleep 10
cd "/var/lib/docker/containers/$SUP_ID"
mv config.v2.json config.v2.json.bak
cat config.v2.json.bak | jq '. | .Path = "/bin/sh"' | jq '. | .Args = ["-c", "sed -i s-current_boot_block_device=.*-current_boot_block_device=/dev/mmcblk0- mount-partitions.sh; exec /usr/src/app/entry.sh"]' | jq -c '.' > config.v2.json
systemctl start balena
echo "balena-supervisor current_boot_partition lsblk fix applied"
```

**Comment by cywang117** [comment]
Thanks for reporting this @jwmullally , I'll take a look at incorporating a fix into the Supervisor.

**Comment by cywang117** [comment]
Resolved with Supervisor v14.10.1.

**Comment by jwmullally** [comment]
Great thanks! 👍 

> `entry.sh` runs with `set -e` then sources `mount-partitions.sh`, which is why the script aborts before the error can be printed.

A separate patch might be needed for this? E.g. also `set -e` in `mount-partitions.sh` so early unplanned exits before error messages can be printed are caught during testing.
